app/controllers/mixins/breadcrumbs_mixin.rb

```diff
@@ -33,33 +34,58 @@ def data_for_breadcrumbs
+
+        # On special page (tagitems, politems, etc.)
+        if @tagitems || @politems || @ownershipitems || @retireitems
+          breadcrumbs.push(special_page_breadcrumb(@tagitems || @politems || @ownershipitems || @retireitems, true)) unless options[:hide_special_item]
+          breadcrumbs.push(:title => @right_cell_text)
+        else
+          # Ancestry parents breadcrumbs (only in services)
+          breadcrumbs.concat(ancestry_parents(options[:ancestry], options[:record_info], options[:record_title])) if options[:ancestry]
 
-        # Append breadcrumbs created from the tree (eg "All policies > Red Hat policies > Policy 1")
-        breadcrumbs_from_tree = build_breadcrumbs_from_tree
-        breadcrumbs.concat(breadcrumbs_from_tree)
+          # Priority 1. Include record
+          #          2. TreeNode.text
+          #          3. Item from tree (if it's built)
+          #          4. Fallbacks (sent text, title of the record, title_for_breadcrumbs)
+          #
+          # Overrides: if x_root show then show title_for_breadcrumbs
 
-        if (options[:include_record] || breadcrumbs_from_tree.blank?) && options[:record_info].present? && options[:record_info][options[:record_title]] &&
-           !(@tagitems || @politems || @ownershipitems || @retireitems)
-          # Append breadcrumb created from record_info if included or if result from tree was empty (eg filters page)
+          # Include items (for trees, which does not have final nodes: VMs etc.)
+          if options[:include_record] && options[:record_info].present?
+            title = options[:record_info][options[:record_title]]
+            key = TreeBuilder.build_node_id(options[:record_info])
+          end
 
-          # Ancestry parents breadcrumbs
-          breadcrumbs.concat(ancestry_parents(options[:ancestry], options[:record_info], options[:record_title])) if options[:ancestry]
+          # Try to create treenode from the x_node
+          entity, entity_id, _ = TreeBuilder.extract_node_model_and_id(x_node) if x_node
+          record = entity.try(:constantize).try(:find, entity_id)
+          title ||= TreeNode.new(record).text if record
 
-          breadcrumbs.push(:title => options[:record_info][options[:record_title]], :key => x_node_right_cell)
-          breadcrumbs.push(:title => title_for_breadcrumbs) if options[:show_header] && @right_cell_text
-        else
-          # Append breadcrumb from the title of right cell
-          breadcrumbs.push(special_page_breadcrumb(@tagitems || @politems || @ownershipitems || @retireitems)) unless options[:hide_special_item]
-          breadcrumbs.push(:title => title_for_breadcrumbs) if action_breadcrumb?
+          # Initial loading of the selected node when coming from different page
+          title ||= build_breadcrumb_from_tree.try(:[], :title) if @trees.present?
+
+          # After switching accordion
+          title = title_for_breadcrumbs if x_node == 'root'
```

Comment:
This is confusing, can we move this above if all the others are with `||=` or wrapped with `if` statements?